lib/msprime.c

```diff
@@ -3169,15 +3169,15 @@ msp_merge_ancestors(msp_t *self, avl_tree_t *Q, population_id_t population_id,
                 }
             }
         } else {
-            if (!coalescence) {
+            if (new_node_id == TSK_NULL) {
                 coalescence = true;
                 l_min = l;
                 ret = msp_store_node(self, 0, self->time, population_id, individual);
                 if (ret != 0) {
                     goto out;
                 }
             }
-            v = (tsk_id_t) msp_get_num_nodes(self) - 1;
+            new_node_id = (tsk_id_t) msp_get_num_nodes(self) - 1;
```

Comment:
This needs to be *inside* the if statement, otherwise we overwrite the supplied new node id value.